File: rust/src/detectors/schema_validation.rs
use crate::detectors::{Definition, Detector, IndexedModel, def_sort_key, make_finding};
use crate::models::{Finding, Severity};
use crate::config::{VariableSchema, SchemaType};
use regex::Regex;

/// Build a validator for a variable schema.
fn build_validator(schema: &VariableSchema) -> Option<Box<dyn Fn(&str) -> Result<(), String> + '_>> {
    if let Some(enum_values) = &schema.enum_values {
        if !enum_values.is_empty() {
            let enum_values = enum_values.clone();
            return Some(Box::new(move |v: &str| {
                if enum_values.contains(&v.to_string()) {
                    Ok(())
                } else {
                    Err(format!("must be one of: {}", enum_values.join(", ")))
                }
            }));
        }
    }

    match schema.var_type {
        Some(SchemaType::Integer) => {
            let min = schema.min;
            let max = schema.max;
            Some(Box::new(move |v: &str| {
                match v.parse::<i64>() {
                    Ok(n) => {
                        if let Some(min) = min {
                            if n < min { return Err(format!("must be >= {}", min)); }
                        }
                        if let Some(max) = max {
                            if n > max { return Err(format!("must be <= {}", max)); }
                        }
                        Ok(())
                    }
                    Err(_) => Err("must be an integer".to_string()),
                }
            }))
        }
        Some(SchemaType::Float) => {
            let min = schema.min;
            let max = schema.max;
            Some(Box::new(move |v: &str| {
                match v.parse::<f64>() {
                    Ok(n) => {
                        if let Some(min) = min {
                            if n < min as f64 { return Err(format!("must be >= {}", min)); }
                        }
                        if let Some(max) = max {
                            if n > max as f64 { return Err(format!("must be <= {}", max)); }
                        }
                        Ok(())
                    }
                    Err(_) => Err("must be a float".to_string()),
                }
            }))
        }
        Some(SchemaType::Boolean) => Some(Box::new(|v: &str| {
            match v.parse::<bool>() {
                Ok(_) => Ok(()),
                Err(_) => Err("must be a boolean".to_string()),
            }
        })),
        Some(SchemaType::Url) => Some(Box::new(|v: &str| {
            if v.starts_with("http://") || v.starts_with("https://") {
                Ok(())
            } else {
                Err("must be a valid URL".to_string())
            }
        })),
        Some(SchemaType::Json) => Some(Box::new(|v: &str| {
            match serde_json::from_str::<serde_json::Value>(v) {
                Ok(_) => Ok(()),
                Err(_) => Err("must be valid JSON".to_string()),
            }
        })),
        Some(SchemaType::Regex) => {
            if let Some(regex_str) = &schema.regex {
                match Regex::new(regex_str) {
                    Ok(re) => Some(Box::new(move |v: &str| {
                        if re.is_match(v) {
                            Ok(())
                        } else {
                            Err(format!("must match {}", regex_str))
                        }
                    })),
                    Err(_) => None,
                }
            } else {
                None
            }
        }
        Some(SchemaType::String) | Some(SchemaType::Enum) | None => Some(Box::new(|_v: &str| Ok(()))),
    }
}

fn validate_value(value: Option<&str>, schema: &VariableSchema) -> Result<(), String> {
    if value.is_none() || value.unwrap().trim().is_empty() {
        if schema.optional.unwrap_or(false) {
            return Ok(());
        }
        return Err("value is required".to_string());
    }

    let validator = build_validator(schema);
    let Some(validator) = validator else {
        return Ok(());
    };

    validator(value.unwrap())
}

/// A variable value does not match its declared schema.
pub struct SchemaValidationDetector;

impl Detector for SchemaValidationDetector {
    fn id(&self) -> &'static str {
        "schema-validation"
    }

    fn name(&self) -> &'static str {
        "schema-validation"
    }

    fn description(&self) -> &'static str {
        "A variable value does not match its declared schema."
    }

    fn detect(&self, index: &IndexedModel) -> Vec<Finding> {
        let schema = &index.model.config.schema;
        if schema.is_empty() {
            return vec![];
        }

        let mut findings = Vec::new();

        let mut entries: Vec<(&String, &Vec<Definition>)> = index.env_definitions.iter().collect();
        entries.sort_by(|(na, da), (nb, db)| def_sort_key(da).cmp(&def_sort_key(db)).then(na.cmp(nb)));

        for (name, defs) in entries {
            let Some(variable_schema) = schema.get(name) else {
                continue;
            };

            for def in defs {
                let result = validate_value(def.value.as_deref(), variable_schema);
                if result.is_ok() {
                    continue;
                }
                findings.push(make_finding(
                    "schema-validation",
                    Severity::Error,
                    name,
                    format!("does not match schema: {}", result.unwrap_err()),
                    vec![def.origin.clone()],
                ));
            }
        }

        findings
    }
}

File: dart/lib/src/detectors/schema_validation.dart
import 'dart:convert';

import '../config/config.dart';
import '../models/finding.dart';
import 'detector.dart';

/// Schema-validation: a variable value does not match its declared schema.
/// Error messages mirror the reference CLI's zod-based validator.
class SchemaValidationDetector implements Detector {
  @override
  String get id => 'schema-validation';

  @override
  String get name => 'schema-validation';

  @override
  String get description =>
      'A variable value does not match its declared schema.';

  @override
  List<Finding> detect(IndexedModel index) {
    final schema = index.model.config.schema;
    if (schema.isEmpty) return [];

    final findings = <Finding>[];

    for (final entry in sortedEntries(index.envDefinitions, defSortKey)) {
      final variableSchema = schema[entry.key];
      if (variableSchema == null) continue;

      for (final def in entry.value) {
        final error = validateValue(def.value, variableSchema);
        if (error == null) continue;
        findings.add(Finding.make(
          'schema-validation',
          Severity.error,
          entry.key,
          'does not match schema: $error',
          [def.origin],
        ));
      }
    }

    return findings;
  }

  static String? validateValue(String? value, VariableSchema schema) {
    if (value == null || value.trim().isEmpty) {
      return schema.optional == true ? null : 'value is required';
    }
    return _buildValidator(schema)?.call(value);
  }

  static String? Function(String)? _buildValidator(VariableSchema schema) {
    final enumValues = schema.enumValues;
    if (enumValues != null && enumValues.isNotEmpty) {
      final quoted = enumValues.map((v) => '"$v"').join('|');
      return (v) =>
          enumValues.contains(v) ? null : 'Invalid option: expected one of $quoted';
    }

    switch (schema.type) {
      case SchemaType.integer:
        final min = schema.min;
        final max = schema.max;
        return (v) {
          final n = jsNumber(v);
          if (n == null || n.isNaN) return 'Invalid input: expected number, received NaN';
          if (!isJsInteger(n)) return 'Invalid input: expected int, received number';
          if (min != null && n < min) {
            return 'Too small: expected number to be >=$min';
          }
          if (max != null && n > max) {
            return 'Too big: expected number to be <=$max';
          }
          return null;
        };
      case SchemaType.float:
        final min = schema.min;
        final max = schema.max;
        return (v) {
          final n = jsNumber(v);
          if (n == null || n.isNaN) return 'Invalid input: expected number, received NaN';
          if (min != null && n < min) {
            return 'Too small: expected number to be >=$min';
          }
          if (max != null && n > max) {
            return 'Too big: expected number to be <=$max';
          }
          return null;
        };
      case SchemaType.boolean:
        // z.coerce.boolean() accepts any non-empty string.
        return (_) => null;
      case SchemaType.url:
        // zod's .url() accepts anything the WHATWG URL constructor accepts.
        return (v) => isValidJsUrl(v) ? null : 'Invalid URL';
      case SchemaType.json:
        return (v) {
          try {
            jsonDecode(v);
            return null;
          } catch (_) {
            return 'must be valid JSON';
          }
        };
      case SchemaType.regex:
        final regexStr = schema.regex;
        if (regexStr == null) return null;
        final RegExp re;
        try {
          re = RegExp(regexStr);
        } catch (_) {
          return null;
        }
        return (v) => re.hasMatch(v) ? null : 'must match $regexStr';
      case SchemaType.string:
      case SchemaType.enumType:
      case null:
        return null;
    }
  }
}

/// JS `Number(value)` for the inputs envdoctor sees: null when NaN.
double? jsNumber(String v) {
  final trimmed = v.trim();
  if (trimmed.isEmpty) return 0; // JS Number('') === 0 (empty values are filtered earlier)
  // Hex / octal literals.
  final hex = RegExp(r'^[+-]?0x[0-9a-fA-F]+$');
  if (hex.hasMatch(trimmed)) {
    final neg = trimmed.startsWith('-');
    final digits = trimmed.replaceFirst(RegExp(r'^[+-]?0x'), '');
    final n = int.tryParse(digits, radix: 16);
    if (n == null) return null;
    return (neg ? -1 : 1) * n.toDouble();
  }
  if (RegExp(r'^[+-]?0o[0-7]+$').hasMatch(trimmed)) {
    final neg = trimmed.startsWith('-');
    final digits = trimmed.replaceFirst(RegExp(r'^[+-]?0o'), '');
    final n = int.tryParse(digits, radix: 8);
    if (n == null) return null;
    return (neg ? -1 : 1) * n.toDouble();
  }
  if (trimmed == 'Infinity' || trimmed == '+Infinity') return double.infinity;
  if (trimmed == '-Infinity') return double.negativeInfinity;
  // Decimal: JS accepts "12", "1.5", ".5", "5.", "1e3", and nothing else.
  if (!RegExp(r'^[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?$').hasMatch(trimmed)) {
    return null; // NaN
  }
  return double.tryParse(trimmed);
}

bool isJsInteger(double n) =>
    n.isFinite && n == n.truncateToDouble() && n.abs() <= 9007199254740991;

/// WHATWG URL validation approximation: a scheme followed by a non-empty
/// hierarchical part, no whitespace.
bool isValidJsUrl(String v) {
  final match = RegExp(r'^([A-Za-z][A-Za-z0-9+.-]*):(.*)$').firstMatch(v.trim());
  if (match == null) return false;
  final rest = match.group(2)!;
  if (rest.isEmpty) return false;
  if (RegExp(r'\s').hasMatch(v.trim())) return false;
  return true;
}
